SalesforceCommerceCloud/pwa-kit
# Responsive images: optimize the catalog images

# Description

Ticket https://gus.lightning.force.com/lightning/r/ADM_Work__c/a07EE00000BP0uXYAT/view

This PR takes advantage of the following items:
- Dynamic Image Service (DIS), which allows you to pass some parameters to easily transform your images
- `sizes` and `srcset` attributes to create the responsive images

With DIS, you can transform with these parameters, for example. For more details, see the [doc](https://documentation.b2c.commercecloud.salesforce.com/DOC1/index.jsp?topic=%2Fcom.demandware.dochelp%2Fcontent%2Fb2c_commerce%2Ftopics%2Fimage_management%2Fb2c_creating_image_transformation_urls.html).
- `q=60`: to compress the image with quality of 60%
- `sw=1000`: to resize the image to be 1000px wide

## `sizes` and `srcset`

If you're not familiar with what `sizes` and `srcset` are, here's a quick intro. `sizes` are like instruction we leave for the browser, while `srcset` is a pool of images with varying widths. The browser chooses the appropriate image width, based on `sizes` we gave them _and_ also the device pixel ratio (most mobile devices have at least 2x density, while 1x is commonly seen on desktop). For more details, see this [MDN page](https://developer.mozilla.org/en-US/docs/Learn/HTML/Multimedia_and_embedding/Responsive_images#resolution_switching_different_sizes).

An example:
```html
<img 
    src="https://example.com/image.jpg" 
    sizes="(min-width: 1200px) 800px, (min-width: 720px) 50vw, 100vw" 
    srcset="https://example.com/image.jpg?sw=800 800w, https://example.com/image.jpg?sw=1600 1600w" />
```

In the above code snippet, the `src` attribute is the fallback for very old browsers who don't support `sizes` or `srcset`. The `sizes` tells the browser _in order_: 
1. if the viewport is larger than 1200px, please use 800px image. 
2. Otherwise, if the viewport is at least 720px wide, select an image whose width is half the viewport's width. 
3. Or else, the image should fill up the whole viewport's width. 

Say, the viewport is currently at 1400px and you're using a mobile device with retina screen (2x density). The browser then would select an image with 800px * 2 = 1600px. It sees that there is a 1600px wide image in the `srcset`. If an exact resolution can't be found, it will choose the closest one.

## Challenges

I was initially not sure what the interface would be like. But inspired from this [Kent C. Dodds article](https://kentcdodds.com/blog/building-an-awesome-image-loading-experience#sizes-srcset-and-cloudinary), I decided to go with a utility function `getImageProps`. Since our page layout design aligns with the default Chakra breakpoints, `getImageProps` allows you to pass in `sizes` that have these breakpoints too.

### Value for `srcset`

Another challenge is what to use for the `srcset` value. Online articles typically gloss over this part. Understandably, it's because it's more like an art than science. It doesn't need to be 100% accurate, but should cover the spectrum of image widths quite well. 

One approach is to use the same set of image widths for every responsive image, while the other approach is more custom and manual. At the end, I decided to go with the latter and use this formula of 5 or so images: mobile 1x, 2x, 3x, and desktop 1x, 2x. Keep in mind too that the more images you provide, the less likely for a cache hit to happen.

### How to manually test

Testing the responsive images turns out to be tricky because of browser caching. Here's what I did to manually test the images:
1. Disable caching in the Network tab
2. Resize the viewport
3. Then reload the page ← if you don't reload, it seems that the browser prefers to use the highest-res image that it has downloaded recently
4. Inspect the image and hover your cursor on it to see what the "current source" is
5. Go to Step 2 and repeat, as you test for different devices

One thing to note is that the images we're working with are actually not great. They're quite pixelated for some reason. To see what I mean, compare these 2 images:
- https://edge.disstg.commercecloud.salesforce.com/dw/image/v2/ZZRF_001/on/demandware.static/-/Sites-apparel-m-catalog/default/dw43751577/images/large/PG.10246370.JJBH3XX.PZ.jpg
- https://edge.disstg.commercecloud.salesforce.com/dw/image/v2/ZZRF_001/on/demandware.static/-/Sites-apparel-m-catalog/default/dw43751577/images/large/PG.10246370.JJBH3XX.PZ.jpg?q=100

Even with or without quality of 100%, the background gradient of both images is not smooth. It looks as if the original images are already compressed too much.

### How much optimization

I found this article that outlines a [more complex optimization](https://jakearchibald.com/2021/serving-sharp-images-to-high-density-screens/) approach. It relies on varying the compression quality value for different device pixel ratio. For 2x images, you can actually use a lower quality value than with a 1x image. So that with 2x images, their file size doesn't balloon too quickly.

But I think the current implementation in this PR is good/simple enough for most developers. 

## Impact

Lighthouse score: I still need to check the lighthouse score. I don't think it'll be a huge change, though. Before the PR, the 'large' images are already at 800px wide, which is not that huge if you consider the high-density screens. Regardless, I predict that having responsive images is still a step forward in the positive direction.

## Remaining todos
- [x] Write tests for the util function
- [x] Look at more pages for other responsive images
- [x] Check lighthouse scores



# Types of Changes




> Breaking changes include:
>
> - Removing a public function or component or prop
> - Adding a required argument to a function
> - Changing the data type of a function parameter or return value
> - Adding a new peer dependency to `package.json`



# Checklists




## General


## Accessibility Compliance

You must check off all items in **one** of the follow two lists:


_or..._


## Localization
